lib/mlir/Dialect/Clift/IR/CliftOps.cpp

```diff
@@ -183,27 +265,419 @@ struct ModuleValidator {
+  Rewriter.eraseOp(Op);
+  return mlir::success();
+}
+
+mlir::LogicalResult MakeLabelOp::verify() {
+  const auto [Assignments, GoTos] = getNumLabelUsers(*this);
+
+  if (Assignments > 1)
+    return emitOpError() << getOperationName()
+                         << " may only have one assignment.";
+
+  if (GoTos != 0 and Assignments == 0)
+    return emitOpError() << getOperationName() << " with a use by "
+                         << GoToOp::getOperationName()
+                         << " must have an assignment.";
+
+  return mlir::success();
+}
+
+//===------------------------------ ReturnOp ------------------------------===//
+
+mlir::LogicalResult ReturnOp::verify() {
+  if (not verifyFunctionReturnType(getExpressionType(getResult())))
+    return emitOpError() << getOperationName()
+                         << " requires void or non-array object type.";
+
+  return mlir::success();
+}
+
+//===------------------------------ SwitchOp ------------------------------===//
+
+void SwitchOp::build(OpBuilder &OdsBuilder,
+                     OperationState &OdsState,
+                     const llvm::ArrayRef<uint64_t> CaseValues) {
+  llvm::SmallVector<int64_t> SignedCaseValues;
+  SignedCaseValues.resize_for_overwrite(CaseValues.size());
+  std::copy(CaseValues.begin(), CaseValues.end(), SignedCaseValues.begin());
+  build(OdsBuilder, OdsState, SignedCaseValues, CaseValues.size());
+}
+
+mlir::ParseResult SwitchOp::parse(OpAsmParser &Parser, OperationState &Result) {
+  // Condition region:
+  Result.addRegion(std::make_unique<Region>());
+
+  // Default case region:
+  Result.addRegion(std::make_unique<Region>());
+
+  if (Parser.parseRegion(*Result.regions[0]).failed())
+    return Parser.emitError(Parser.getCurrentLocation(),
+                            "Expected switch condition region");
+
+  llvm::SmallVector<int64_t, 16> CaseValues;
+  while (Parser.parseOptionalKeyword("case").succeeded()) {
+    uint64_t CaseValue;
+    if (Parser.parseInteger(CaseValue).failed())
+      return Parser.emitError(Parser.getCurrentLocation(),
+                              "Expected switch case value");
+
+    auto R = std::make_unique<Region>();
+    if (Parser.parseRegion(*R).failed())
+      return Parser.emitError(Parser.getCurrentLocation(),
+                              "Expected switch case region");
+
+    CaseValues.push_back(static_cast<uint64_t>(CaseValue));
+    Result.addRegion(std::move(R));
+  }
+
+  if (Parser.parseOptionalKeyword("default").succeeded()) {
+    if (Parser.parseRegion(*Result.regions[1]).failed())
+      return Parser.emitError(Parser.getCurrentLocation(),
+                              "Expected switch default region");
+  }
+
+  Result.attributes.set("case_values",
+                        DenseI64ArrayAttr::get(Parser.getContext(),
+                                               CaseValues));
+
+  if (Parser.parseOptionalAttrDict(Result.attributes).failed())
     return mlir::failure();
+
+  return mlir::success();
```

Comment:
For parsing, let me see if I understand what's happening in these last few lines.

`Result.attributes.set("case_values", ...)` is just registering in the `Result` the `CaseValues` that it has parsed above in the `while` loop.

An then what's the catch with `Parser.ParseOptionalAttrDict`?
My understanding is that we don't allow any optional attribute on the `clift::SwitchOp`, so I expect that to always fail.
Or maybe it only succeeds when there's nothing left to parse, which means this is a way to detect when someone is passing additional unexpected attributes. If this is the case then I' happy and feel free to close this thread.

Reply:
You can always add arbitrary named attributes to operations. They don't all have to be declared in ODS. Users can use these attributes to attach their own data to operations. One example of how *we* are currently using this is attaching model metadata to LLVMIR function operations under the name `revng.function.entry`.

Reply:
Oh ok I tend to forget this.

Thanks